src/csharp/Grpc.Core.Tests/Internal/AsyncCallTest.cs

```diff
@@ -472,6 +472,113 @@ public void ServerStreaming_MoreResponses_Success()
             AssertStreamingResponseSuccess(asyncCall, fakeCall, readTask3);
         }
 
+        [Test]
+        public void ServerStreaming_CancelCall()
+        {
+            // Test client cancelling the call before the reading the server stream is complete.
+
+            asyncCall.StartServerStreamingCall("request1");
+            var responseStream = new ClientResponseStream<string, string>(asyncCall);
+
+            // Check that the client has a reference to exactly one call
+            Assert.AreEqual(1, channel.GetCallReferenceCount());
+            // Check that the call is not yet disposed
+            Assert.IsFalse(fakeCall.IsDisposed);
+
+            fakeCall.ReceivedResponseHeadersCallback.OnReceivedResponseHeaders(true, new Metadata());
+            Assert.AreEqual(0, asyncCall.ResponseHeadersAsync.Result.Count);
+
+            // read some data
+            var readTask1 = responseStream.MoveNext();
+            fakeCall.ReceivedMessageCallback.OnReceivedMessage(true, CreateResponsePayload());
+            Assert.IsTrue(readTask1.Result);
+            Assert.AreEqual("response1", responseStream.Current);
+
+            // Cancel the call. This will trigger ReceivedStatusOnClientCallback to be called with a Cancelled status.
+            asyncCall.Cancel();
+            fakeCall.ReceivedStatusOnClientCallback.OnReceivedStatusOnClient(true, new ClientSideStatus(Status.DefaultCancelled, new Metadata()));
+
+            // Check that the resources have been released and that the client no longer references the call
+            Assert.AreEqual(0, channel.GetCallReferenceCount());
+            Assert.IsTrue(fakeCall.IsDisposed);
+        }
+
+        [Test]
+        public void ServerStreaming_CancelCallMoreResponses()
+        {
+            // Test client cancelling the call before the reading the server stream is complete.
+            // This version of the test simulates a race condition where the some data is
+            // received before the OnReceivedStatusOnClient callback is triggered by the cancel.
+
+            asyncCall.StartServerStreamingCall("request1");
+            var responseStream = new ClientResponseStream<string, string>(asyncCall);
+
+            Assert.AreEqual(1, channel.GetCallReferenceCount());
+            Assert.IsFalse(fakeCall.IsDisposed);
+
+            fakeCall.ReceivedResponseHeadersCallback.OnReceivedResponseHeaders(true, new Metadata());
+            Assert.AreEqual(0, asyncCall.ResponseHeadersAsync.Result.Count);
+
+            var readTask1 = responseStream.MoveNext();
+            fakeCall.ReceivedMessageCallback.OnReceivedMessage(true, CreateResponsePayload());
+            Assert.IsTrue(readTask1.Result);
+            Assert.AreEqual("response1", responseStream.Current);
+
+            // Cancel the call
+            asyncCall.Cancel();
```

Comment:
I thought the idea here would be to test an "inflight" read
(MoveNext has been called, but the OnReceivedMessage callback hasn't been called yet when OnReceivedStatusOnClient is received)
Generally speaking the OnReceivedStatusOnClient callback can be delivered earlier than the last (or last few) OnReceivedMessage callbacks, since there can be a race between delivering the events on two different threads in the threadpool.

So I think you might wanna add one more test to test the "inflight" read scenario.